src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -125,5 +147,110 @@ public static partial class JsonSerializer
+            }
+
+            MetadataPropertyName metadata = GetMetadataPropertyName(propertyName, ref state, ref reader);
+            state.Current.MetadataProperty = metadata;
+            return metadata;
+        }
+
+        private static void ResolveMetadataOnDictionary(MetadataPropertyName metadata, ref ReadStack state)
+        {
+            bool isPreserved = state.Current.IsProcessingProperty(ClassType.Dictionary) ?
+                state.Current.DictionaryPropertyIsPreserved : state.Current.IsPreserved;
+
+            if (metadata == MetadataPropertyName.Id)
+            {
+                // Check we are not parsing into immutable.
+                if (state.Current.TempDictionaryValues != null)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(state.Current.JsonPropertyInfo.DeclaredPropertyType);
+                }
+
+                if (state.Current.KeyName != null || isPreserved)
+                {
+                    state.Current.KeyName = null;
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                }
+
+                SetAsPreserved(ref state.Current);
+                state.Current.ReadMetadataValue = true;
+            }
+            else if (metadata == MetadataPropertyName.Ref)
+            {
+                if (state.Current.KeyName != null || isPreserved)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties();
+                }
+
+                state.Current.ShouldHandleReference = true;
+                state.Current.ReadMetadataValue = true;
+            }
+        }
+
+        private static void ResolveMetadataOnObject(ReadOnlySpan<byte> propertyName, MetadataPropertyName meta, ref ReadStack state, ref Utf8JsonReader reader, JsonSerializerOptions options)
+        {
+            if (meta == MetadataPropertyName.NoMetadata)
+            {
+                if (state.Current.IsPreservedArray)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(state.Current.JsonClassInfo.PropertyCache["Values"].DeclaredPropertyType);
+                }
+                // Regular property, call main logic for HandlePropertyName.
+                HandlePropertyNameDefault(propertyName, ref state, ref reader, options);
+            }
+            else if (meta == MetadataPropertyName.Id)
+            {
+                if (state.Current.PropertyIndex > 0 || state.Current.IsPreserved)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                }
+
+                JsonPropertyInfo info = JsonPropertyInfo.s_metadataProperty;
+                info.JsonPropertyName = propertyName.ToArray();
+                state.Current.JsonPropertyInfo = info;
```

Comment:
same as in https://github.com/dotnet/runtime/pull/655/files#r363527332, to build up the exception's JSON path in case something goes wrong later on the object graph.